### DISCUSSION TITLE
Global import

### DISCUSSION THREAD

**Discussion by jongbonga** [ng-icons/ng-icons#141] [Q&A] [answered]
I have `provideIcons(hugeIcons)` in `app.module` but when I call <ng-icon name="hugeMushroom"></ng-icon> in components, nothing displays. 
Is there a way to not provide or import it everywhere? I'm not using standalone components

**Comment by ashley-hunter** [comment, marked as answer]
Hi! Sure, so if you're not using standalone components you should try using the `withIcons` function on `NgIconsModule` instead of `provideIcons` which may resolve this issue:

```ts
import * as hugeIcons from '@ng-icons/huge-icons';

@NgModule({
   imports: [
      NgIconsModule.withIcons(hugeIcons),
   ]
})
export class AppModule {}
```

Note that importing all icons has the downside of bloating your bundle size, whereas importing them individually, while a tad more cumbersome, is much more optimized.

Here is an example of importing all icons in AppModule:

https://stackblitz.com/edit/stackblitz-starters-cuv3s9?description=An%20angular-cli%20project%20based%20on%20@angular/animations,%20@angular/common,%20@angular/compiler,%20@angular/core,%20@angular/forms,%20@angular/platform-browser,%20@angular/platform-browser-dynamic,%20@angular/router,%20core-js,%20rxjs,%20tslib%20and%20zone.js&file=src/main.ts,src/icon-list.ts&template=node&title=Angular%20Starter

If you have any issues, if you are able to reproduce the issue I'd be happy to take a look, thank you!

**Reply by jongbonga** [reply]
Thank you for getting back to me so quickly. I will try and revert to you. 
Is there no way to tree shake what's not being used besides manual import?

**Reply by ashley-hunter** [reply]
No problem!

> Is there no way to tree shake what's not being used besides manual import?

Not really, you are bringing in all the icons in the root of your application, and registering them - as far as a bundler can tell you have imported everything from that library and are using it so it can't safely remove any icons as you could use them at some point in your application.

This is really the compromise, you either load all icons - the benefit is easier dev experience, but larger app size and slower load time, or you selectively import icons, slightly more work for the dev to do, but much smaller app size and faster load times.

You have a few options to improve things:

1. Only register icons that you know you are using in your app, registering them once in the AppModule will make them accessible to every component in your application, so you only have to register them once. You still have to selectively import them, but only once in the root of your app.

2. Use a dynamic loader (https://github.com/ng-icons/ng-icons?tab=readme-ov-file#dynamically-loading-icons). While this will not reduce the overall bundle size, what it will do is lazy load the iconset so it is non-blocking, so actual app loading times will remain reduced, but icons may appear blank until they have loaded.

Something like this should work (I haven't tested this code so it may need slightly tweaked):

```ts
provideNgIconLoader(async name => {
   const icons = await import('@ng-icons/huge-icons');
   return icons[name];
}),
```

**Reply by jongbonga** [reply]
Thank you, I really appreciate your time and effort

**Reply by ashley-hunter** [reply]
No problem! Thank you!
